elements/castbar.lua

```diff
@@ -83,462 +85,289 @@ A default texture will be applied to the StatusBar and Texture widgets if they d
+	if(self.unit ~= unit) then return end
 
 	local element = self.Castbar
-	if(element.castID ~= castID) then
+	if(not element:IsShown() or element.castID ~= castID or element.spellID ~= spellID) then
 		return
 	end
 
-	element.casting = nil
-	element.notInterruptible = nil
+	resetAttributes(element)
 
-	--[[ Callback: Castbar:PostCastStop(unit)
-	Called after the element has been updated when a spell cast has finished.
+	--[[ Callback: Castbar:PostCastStop(unit, spellID)
+	Called after the element has been updated when a spell cast has stopped.
 
-	* self - the Castbar widget
-	* unit - unit for which the update has been triggered (string)
+	* self    - the Castbar widget
+	* unit    - the unit for which the update has been triggered (string)
+	* spellID - the ID of the spell (number)
 	--]]
 	if(element.PostCastStop) then
-		return element:PostCastStop(unit)
+		return element:PostCastStop(unit, spellID)
 	end
 end
 
-local function UNIT_SPELLCAST_CHANNEL_START(self, event, unit, _, spellID)
-	if(self.unit ~= unit and self.realUnit ~= unit) then return end
+local function CastFail(self, event, unit, castID, spellID)
+	if(self.unit ~= unit) then return end
 
 	local element = self.Castbar
-	local name, _, texture, startTime, endTime, _, notInterruptible = UnitChannelInfo(unit)
-	if(not name) then
+	if(not element:IsShown() or element.castID ~= castID or element.spellID ~= spellID) then
 		return
 	end
 
-	endTime = endTime / 1e3
-	startTime = startTime / 1e3
-	local max = (endTime - startTime)
-	local duration = endTime - GetTime()
-
-	element.duration = duration
-	element.max = max
-	element.delay = 0
-	element.channeling = true
-	element.notInterruptible = notInterruptible
-	element.holdTime = 0
-	element.spellID = spellID
-
-	-- We have to do this, as it's possible for spell casts to never have _STOP
-	-- executed or be fully completed by the OnUpdate handler before CHANNEL_START
-	-- is called.
-	element.casting = nil
-	element.castID = nil
-
-	element:SetMinMaxValues(0, max)
-	element:SetValue(duration)
+	if(element.Text) then
+		element.Text:SetText(event == 'UNIT_SPELLCAST_FAILED' and FAILED or INTERRUPTED)
+	end
 
-	if(element.Text) then element.Text:SetText(name) end
-	if(element.Icon) then element.Icon:SetTexture(texture or FALLBACK_ICON) end
-	if(element.Time) then element.Time:SetText() end
+	if(element.Spark) then element.Spark:Hide() end
 
-	local shield = element.Shield
-	if(shield and notInterruptible) then
-		shield:Show()
-	elseif(shield) then
-		shield:Hide()
-	end
+	element.holdTime = element.timeToHold or 0
 
-	if(element.SafeZone) then
-		positionSafeZone(element, true)
-		updateSafeZone(element)
-	end
+	resetAttributes(element)
+	element:SetValue(element.max)
 
-	--[[ Callback: Castbar:PostChannelStart(unit, name)
-	Called after the element has been updated upon a spell channel start.
+	--[[ Callback: Castbar:PostCastFail(unit, spellID)
+	Called after the element has been updated upon a failed spell cast.
```

Comment:
Called after the element has been updated upon a failed**/interrupted** spell cast.